arbiter/drivers/s3.cpp

```diff
@@ -223,9 +223,55 @@ std::unique_ptr<S3::Auth> S3::Auth::create(
     drivers::Http httpDriver(pool);
 
     // Nothing found in the environment or on the filesystem.  However we may
-    // be running in an EC2 instance with an instance profile set up.
+    // be running in an EC2 instance with a service account or an instance profile set up.
     try
     {
+        // Try to "assume role with web identity" - see https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html
+        try
+        {
+            const auto roleArn = env("AWS_ROLE_ARN");
+            const auto webIdentityTokenFile = env("AWS_WEB_IDENTITY_TOKEN_FILE");
+            std::unique_ptr<std::string> webIdentityToken;
+
+            if (roleArn && webIdentityTokenFile && (webIdentityToken = fsDriver.tryGet(*webIdentityTokenFile)))
+            {
+                // Decide on the STS root URL, defaults to https://sts.<AWS_REGION>.amazonaws.com
+                std::string stsRootUrl;
+                if (env("AWS_STS_ROOT_URL"))
+                {
+                    stsRootUrl = *env("AWS_STS_ROOT_URL");
+                }
+                else
+                {
+                    bool useRegionalEndpoint = env("AWS_STS_REGIONAL_ENDPOINTS")
+                        ? (*env("AWS_STS_REGIONAL_ENDPOINTS") == "regional")
+                        : true;
+                    if (useRegionalEndpoint)
+                    {
+                        stsRootUrl = "https://sts." + S3::Config::extractRegion(s, profile) + ".amazonaws.com";
+                    }
+                    else
+                    {
+                        stsRootUrl = "https://sts.amazonaws.com";
+                    }
+                }
+
+                const std::string stsAssumeRoleWithWebIdentityUrl = stsRootUrl
+                    + "/?Action=AssumeRoleWithWebIdentity&RoleSessionName=pdal&Version=2011-06-15"
```

Comment:
Should `RoleSessionName=pdal` be a configurable parameter?

Reply:
Surely someone will appreciate it being configurable

Done